### PR TITLE
[AutoDiff] Revamp differentiation transform.

### PR DESCRIPTION
**Pull request by dan-zheng**
- Directly generate primal code in VJP functions.
  - Rename `PrimalGenCloner` to `VJPEmitter`.
  - This unblocks control-flow support.
    Primal data structures can be generated based on activity analysis.
- Refactor differentiation transform to use iterative `autodiff_function`
  instruction worklist instead of a fixed `DifferentiationTask` worklist.
  - `VJPEmitter::visitApply` now creates `autodiff_function` and
    `autodiff_function_extract` instructions instead of directly emitting
    associated function references.
  - An iterative loop canonicalizes `autodiff_function` instructions,
    promoting them to `@differentiable` function-typed values.
- Add `autodiff_function_extract` folding optimization.
  - Fold `autodiff_function_extract` users of `autodiff_function` instructions,
    directly replacing them with operands of the `autodiff_function`
    instruction.
  - If the `autodiff_function` instruction has only `autodiff_function_extract`
    users, delete the instruction itself after folding.
- Remove unnecessary auxiliary data structures.
  - `DifferentiationTask`: replace with map from `[differentiable]` attributes
    to `DifferentiationInvoker`.
  - `PrimalGen`, `AdjointGen`: replace with `autodiff_function` iterative loop.

### CODE REVIEW

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: lib/SILOptimizer/Utils/Generics.cpp
Comment (dan-zheng) [inline lib/SILOptimizer/Utils/Generics.cpp]: This change is necessary to prevent post-differentiation SIL passes from generating invalid SIL.

---

In particular, this change disables specialization of operands to `autodiff_function`.
It may be possible to teach the specializer how to handle `autodiff_function` instructions. I don't know much work that would take, it seems difficult.

File: lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp
Comment (dan-zheng) [inline lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp]: This change is necessary to prevent post-differentiation SIL passes from generating invalid SIL.

---

This change is rather innocuous and can be upstreamed to `master`. It just adds a previously unexercised check.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Todo: add documentation.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Todo: remove stray comment.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Note: indirect differentiation invokers are currently not handled post-refactoring. I've disabled the test for now.

File: lib/SILOptimizer/IPO/CapturePropagation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/IPO/CapturePropagation.cpp]: Yes, I also believe so.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: I deleted that line to unblock compilation errors I wasn't sure how to fix.

Here are the errors when I add the line back:
```
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:3545:19: error: call to deleted constructor of '(anonymous namespace)::PrimalInfo'
        vjp(vjp), primalInfo(primalInfo), invoker(invoker),
                  ^          ~~~~~~~~~~
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:432:3: note: 'PrimalInfo' has been explicitly marked deleted here
  PrimalInfo(const PrimalInfo &) = delete;
  ^
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:5248:33: error: call to deleted constructor of '(anonymous namespace)::PrimalInfo'
                                *primalInfo.get(), invoker, activityInfo,
                                ^~~~~~~~~~~~~~~~~
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:432:3: note: 'PrimalInfo' has been explicitly marked deleted here
  PrimalInfo(const PrimalInfo &) = delete;
  ^
/Users/danielzheng/swift-dev/swift/lib/SILOptimizer/Mandatory/Differentiation.cpp:3540:38: note: passing argument to parameter 'primalInfo' here
                          PrimalInfo primalInfo,
                                     ^
```

The two copies happen in:
* The constructor of `AdjointEmitter`.
* The call site of the constructor of `AdjointEmitter` (`*primalInfo.get()`).

Do you have any suggestions?

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: `generatedFunctionReferences` stores the `SILValue` outputs of `emitAssociatedFunctionReference`:
```
getGeneratedFunctionReferences().push_back(assocFn);
```

So `FunctionRefInst *` is not appropriate as the element type. I'll leave things as is.

Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: The doc comment is not accurate, I'll try to improve it.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: When you take it as an argument, it should be `PrimalInfo &primalInfo` instead of `PrimalInfo primalInfo`.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Improved doc comment and naming:
```
  /// List of associated function references, generated via
  /// `emitAssociatedFunctionReference`.
  /// Saved for deletion during cleanup.
  SmallVector<SILValue, 32> generatedAssociatedFunctionReferences;
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Good point, I'll remove the blurb.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: We discussed in person. The semantics of the method are a bit complicated.
We agreed to use the original name of this method, `promoteToDifferentiableFunction`, since the method always returns a `SILValue` with `@differentiable` function type.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: That works, thanks!

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Sorry, I accidentally copied the old doc comment `Queue of differentiation tasks.` without editing.

`autoDiffFunctionInsts` is definitely a stack (processed via while loop and `pop_back`). I'll update the doc comment:
```cpp
  /// The worklist (stack) of `autodiff_function` instructions to be processed.
  SmallVector<AutoDiffFunctionInst *, 32> autoDiffFunctionInsts;
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: We discussed in person. You had the idea of effectively setting `AutoDiffFunctionInst` to nullptr in-place, which I've implemented.
```
  // Delete all worklist occurrences of `adfi` by setting them to nullptr.
  // This is more efficient than APIs like `llvm::erase_if`.
  for (auto &inst : getAutoDiffFunctionInsts())
    if (inst == adfi)
      inst = nullptr;
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: The triple for indirect differentiation is ``(original `apply` instruction, original function, `[differentiable]` attribute)`` for the parent invoker - the original function and `[differentiable]` attribute can be used to index into `ADContext::invokers` to get the parent invoker.

Your comment made me realize storing the original function is redundant, because the original function can be simply retrieved from the `apply` instruction. Does storing a `std::pair<ApplyInst *, SILDifferentiableAttr *>` (for now) make sense to you?

Comment (rxwei) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: FWIW, `SILDifferentiableAttr` actually stores the original function.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: `desiredIndices` and `actualIndices` should be removed from `NestedApplyInfo`, now that `emitAssociatedFunctionReference` is no longer called within `visitApply`.

`actualIndices` are resolved in `ADContext::processAutoDiffFunctionInst`.

Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Done in https://github.com/apple/swift/pull/24845/commits/10522e5882b8c1e0714a9523827abc2a653a0187. Also removed original `SILFunction *` from `SILDifferentiableAttribute` `DifferentiationInvoker` case.

File: test/AutoDiff/autodiff_indirect_diagnostics.swift
Comment (dan-zheng) [reply test/AutoDiff/autodiff_indirect_diagnostics.swift]: I'm actually not sure what's the expected behavior for these tests. I'll debug a bit to see what's going on.

At a glance, the changes seem unideal. The old diagnostics were more informative and appeared on the `gradient` user invocation.
```diff
--- a/test/AutoDiff/autodiff_indirect_diagnostics.swift
+++ b/test/AutoDiff/autodiff_indirect_diagnostics.swift
@@ -6,14 +6,13 @@

 // Test unmet generic requirements.

-// expected-error @+1 {{function is not differentiable}}
 @differentiable
-// expected-note @+1 {{when differentiating this function definition}}
 func generic<T: Differentiable & FloatingPoint>(_ x: T) -> T {
+  // expected-error @+2 {{expression is not differentiable}}
   // expected-note @+1 {{member is not differentiable because the corresponding protocol requirement is not '@differentiable'}}
   return x + 1
 }
-_ = gradient(at: 1.0, in: generic) // expected-error {{function is not differentiable}}
+_ = gradient(at: 1.0, in: generic)
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Please make case labels align with `switch`.

**Comment by dan-zheng** [comment]
Ready for re-review. cc @rxwei 
I'm investigating why diagnostics changed in `test/AutoDiff/autodiff_indirect_diagnostics.swift`.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: This logic should be unified with existing logic used for `emitZeroIndirect` instead of being duplicated.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
        SmallVector<SILValue, 8> arguments(ai->getArguments().begin(),
                                           ai->getArguments().end())
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: Maybe an early return here?

```suggestion
    if (!adfi) continue;
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]: In SIL differentiation, "index subset" can mean both parameter indices and result indices. This name is a bit unclear to me.

```suggestion
  getOrCreateSubsetParametersThunkForAssociatedFunction(
```

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (rxwei) [inline lib/SILOptimizer/Mandatory/Differentiation.cpp]:
```suggestion
  SILFunction *getOrCreateSubsetParametersThunkForLinearMap(
```

**Review by rxwei — APPROVED** [review]
The patch LGTM overall. Huge congrats on getting this done!

There are two necessary improvements I think should be done before merging:
1. The `AdditiveArithmetic.zero` emission logic in a thunk builder should be unified with what's already in `AdjointEmitter`.
2. I suspect there's a performance regression caused by extra `autodiff_function` and `autodiff_function_extract` instructions being generated, which could pose a barrier for later optimizations. Folding trivial `autodiff_function`-`autodiff_function_extract` can be trivially done near the logic that canonicalizes `autodiff_function` instructions in the main loop.

Other than those, we should talk about fixing diagnostics for indirect calls soon.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Oh nice. I've been using the suboptimal `map<SmallVector<SILValue, ...>>` pattern in multiple places.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: I'm a fan of `if (!adfi) continue;` just for the uniform indentation. 😃

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Thanks! `IndexSubset` was definitely a suboptimal name.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Done. Also changed another `map<SmallVector<...>>` call-site to `SmallVector` direct constructor call.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Done in https://github.com/apple/swift/pull/24845/commits/646096e004e5525b1baefd17701bd4f07a55ad40.

File: lib/SILOptimizer/Mandatory/Differentiation.cpp
Comment (dan-zheng) [reply lib/SILOptimizer/Mandatory/Differentiation.cpp]: Done.

**Comment by dan-zheng** [comment]
Thanks @rxwei for the thorough review!

1. Unify `zero` emission logic is done in https://github.com/apple/swift/pull/24845/commits/646096e004e5525b1baefd17701bd4f07a55ad40.
2. `autodiff_function_extract` folding optimization is done in https://github.com/apple/swift/pull/24845/commits/064e61e6bf1c76b5d1813c8cec552bb0bd5b9f6e.
    * As we discussed, a `-Xllvm -differentiation-skip-folding-autodiff-function-extraction` flag was added to disable folding for SIL testing purposes. Used in three tests.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

File: test/AutoDiff/autodiff_indirect_diagnostics.swift
Comment (dan-zheng) [reply test/AutoDiff/autodiff_indirect_diagnostics.swift]: I didn't quite make sense of the diagnostic behavior changes yet.
We plan to change/fix diagnostic behavior soon, so it would be good to revisit these changes as part of that follow-up.

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Comment by dan-zheng** [comment]
@swift-ci Please test tensorflow

**Pull request merged**
